test/OtokenFactory.test.ts

```diff
@@ -0,0 +1,221 @@
+import {
+  OtokenFactoryInstance,
+  MockOtokenInstance,
+  MockAddressBookInstance,
+  MockWhitelistModuleInstance,
+  MockERC20Instance,
+} from '../build/types/truffle-types'
+import BigNumber from 'bignumber.js'
+import {assert} from 'chai'
+const {expectEvent, expectRevert, time} = require('@openzeppelin/test-helpers')
+const OTokenFactory = artifacts.require('OtokenFactory.sol')
+const MockAddressBook = artifacts.require('MockAddressBook.sol')
+const MockWhitelist = artifacts.require('MockWhitelistModule.sol')
+const MockERC20 = artifacts.require('MockERC20.sol')
+const MockOtoken = artifacts.require('MockOtoken.sol')
+const ZERO_ADDR = '0x0000000000000000000000000000000000000000'
+
+contract('OTokenFactory', accounts => {
+  let otoken: MockOtokenInstance
+  let addressBook: MockAddressBookInstance
+  let otokenFactory: OtokenFactoryInstance
+
+  // Paramter used for otoken init(). (Use random addresses as usdc and eth)
+  let usdc: MockERC20Instance
+  let shitcoin: MockERC20Instance
+  const ethAddress = ZERO_ADDR
+  const strikePrice = new BigNumber(200).times(new BigNumber(10).exponentiatedBy(18))
+  const isPut = true
+  let expiry: number
+
+  before('Deploy otoken logic and Factory contract', async () => {
+    usdc = await MockERC20.new('USDC', 'USDC')
+    shitcoin = await MockERC20.new('Shit coin', 'STC')
+
+    const logic = await MockOtoken.new()
+
+    // Deploy and whitelist ETH:USDC product
+    const mockWhitelist: MockWhitelistModuleInstance = await MockWhitelist.new()
+    await mockWhitelist.whitelistProduct(ethAddress, usdc.address, usdc.address)
+    // Deploy addressbook
+    addressBook = await MockAddressBook.new()
+    await addressBook.setOtokenImpl(logic.address)
+    await addressBook.setWhitelist(mockWhitelist.address)
+
+    otokenFactory = await OTokenFactory.new(addressBook.address)
+    expiry = (await time.latest()).toNumber() + time.duration.days(30).toNumber()
+  })
+
+  describe('Get otoken address', () => {
+    it('Should have no otoken records at the begining', async () => {
+      const otokens = await otokenFactory.getOtokens()
+      assert.equal(otokens.length, 0, 'Should have no otoken records')
+    })
+
+    it('Should return address(0) if token is not deployed', async () => {
+      const existAddress = await otokenFactory.getOtoken(
+        ethAddress,
+        usdc.address,
+        usdc.address,
+        strikePrice,
+        expiry,
+        isPut,
+      )
+      assert.equal(existAddress, ZERO_ADDR, 'getOtoken check failed on undeployed tokens.')
+    })
+
+    it('should get deterministic address with new otoken paramters', async () => {
+      const targetAddress = await otokenFactory.getTargetOtokenAddress(
```

Comment:
can you also test that calling this multiple times should give the same address all times?

Reply:
added